text/0000-router-service-computed-query-params.md

```diff
@@ -0,0 +1,324 @@
+    serialize(queryParams: object): string {
+      // serialize object for query string
+      // default to URLSearchParams, polyfilled for IE11
+    },
+    deserialize(queryString: string): object {
+      // parse to object for use in `injectedRouter.queryString`
+      // also default to URLSerachParams
+    }
+  }
+});
+```
+
+
+This will address a long standing issues from as far back as 2016,
+some new functionality for serialization and deserialization could be powered by [qs](https://www.npmjs.com/package/qs) ([3.4kb (gzip+min)](https://bundlephobia.com/result?p=qs@6.7.0)) or a lternatively, [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) -- this would enable the setting of arrays and objects which is not possible today.
+
+
+### Sticky Query Params
+
+By default, transitionTo will clear the query params, unless specified inside the transiion.
+
+If query params are defined ahead of time as sticky, they will persist in the URL between sub routes.
+
+This can be configured in the `Router.map` function:
+```ts
+Router.map(function() {
+  this.queryParams('bar');
+
+  this.route('faq');
+
+  this.route('posts', function() {
+    this.queryParams('foo', 'baz');
+
+    this.route('new');
+    this.route('index');
+    this.route(
+      'show',
+      { path: '/:postId', queryParams: ['hideComments', 'invertColors'] },
+      function() {
+        this.route('edit');
+        this.route('comment');
+        this.route('share');
+      }
+    );
+  });
+});
+```
+
+This method of dealing with query params implies the following:
+ - All query params, even if not specified are allowed. given the above example, I could use the qp "strongestAvenger" anywhere
+ - Any sort of transition will clear the query params, unless it is defined as a sticky queryParam. So, if I'm on the posts/new route with the query params "foo" and "baz", and transition to posts/show, those query params are still available. If I navigate to the faq page, foo and baz are cleared from the URL.
+ - The globally defined query params will stick around until cleared manually. If I visit faq?bar=2, and then transition to posts. the bar=2 query param will still be present.
+ - The `this.queryParams` function will be available at every nesting of the route tree, but `queryParams` are also configurable in the route options hash.
+
+Notes:
+ - This does not imply a caching mechanism (like what controllers today allow)
+ - If a certain app wants caching of query params as they exist today, an addon could be made to fill that gap.
+
+ **A more concrete example: filter query params**
+Given we want a way to search over a list of products, be able to view additional information about a selected search result, and save the search result for later, sticky params help maintain the search throughout all of the route navigations (similar to Amazon). 
+
+
+```ts
+Router.map(function() {
+  this.queryParams('bar');
+
+  this.route('search', { queryParams: [
```

Comment:
Lets format this with prettier (and the other example snippets), I think it would end up something like:

```js
this.route(
  'search',
  {
    // list here
  },
  function() {

  }
);
```